ccl/changefeedccl: TestChangefeedCursor failed

ccl/changefeedccl.TestChangefeedCursor [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10138140?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10138140?buildTab=artifacts#/) on release-23.1 @ [4d53217fb2d27559257d46261badb3be91428d1a](https://github.com/cockroachdb/cockroach/commits/4d53217fb2d27559257d46261badb3be91428d1a):


```
=== RUN   TestChangefeedCursor
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/905465f697dc04c39f7d8a82cb057bfe/logTestChangefeedCursor1174416622
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestChangefeedCursor
    changefeed_test.go:879: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/905465f697dc04c39f7d8a82cb057bfe/logTestChangefeedCursor1174416622
--- FAIL: TestChangefeedCursor (314.23s)
=== RUN   TestChangefeedCursor/enterprise
    helpers_test.go:820: making server as system tenant
    helpers_test.go:895: making enterprise feed factory
    helpers_test.go:909: pgURL enterprise root
    changefeed_test.go:861: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:195
        	            				github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:270
        	            				github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_test.go:861
        	            				github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:959
        	            				github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:990
        	Error:      	Received unexpected error:
        	            	operation "withTimeout-866120114506530817" timed out after 5m0s (given timeout 5m0s): operation "tableFeed.Next-866120114506530817" timed out after 5m0s (given timeout 5m0s): context deadline exceeded
        	            	(1) operation "withTimeout-866120114506530817" timed out after 5m0s (given timeout 5m0s)
        	            	Wraps: (2) operation "tableFeed.Next-866120114506530817" timed out after 5m0s (given timeout 5m0s)
        	            	Wraps: (3) context deadline exceeded
        	            	Error types: (1) *contextutil.TimeoutError (2) *contextutil.TimeoutError (3) context.deadlineExceededError
        	Test:       	TestChangefeedCursor/enterprise
    --- FAIL: TestChangefeedCursor/enterprise (314.22s)
```
<p>Parameters: <code>TAGS=bazel,gss,race</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #97317 ccl/changefeedccl: TestChangefeedCursor failed [C-test-failure O-robot branch-master]
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedCursor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-28077

--- a/pkg/ccl/changefeedccl/changefeed_test.go
+++ b/pkg/ccl/changefeedccl/changefeed_test.go
@@ -72,6 +72,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
 	"github.com/cockroachdb/cockroach/pkg/sql/randgen"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/eval"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/storage"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
@@ -792,21 +793,23 @@ func TestChangefeedCursor(t *testing.T) {
 		sqlDB := sqlutils.MakeSQLRunner(s.DB)
 		sqlDB.Exec(t, `CREATE TABLE foo (a INT PRIMARY KEY, b STRING)`)
 
-		// To make sure that these timestamps are after 'before' and before
-		// 'after', throw a couple sleeps around them. We round timestamps to
-		// Microsecond granularity for Postgres compatibility, so make the
-		// sleeps 10x that.
+		// NB: The test server is a single node and the hlc clock is a
+		// singleton. Any transaction (ie. `INSERT INTO`) or call to
+		// s.Server.Clock().Now() will share this clock. Any read of the clock
+		// increments its current logical time. Thus, the operations below which
+		// happen in sequence will have strictly increasing logical timestamps.
 		beforeInsert := s.Server.Clock().Now()
 		sqlDB.Exec(t, `INSERT INTO foo VALUES (1, 'before')`)
-		time.Sleep(10 * time.Microsecond)
+		insertTimestamp := s.Server.Clock().Now()
 
-		var tsLogical string
-		sqlDB.QueryRow(t, `SELECT cluster_logical_timestamp()`).Scan(&tsLogical)
-		var tsClock time.Time
-		sqlDB.QueryRow(t, `SELECT clock_timestamp()`).Scan(&tsClock)
+		tsLogical := s.Server.Clock().Now()
+		tsClock := timeutil.FromUnixNanos(tsLogical.WallTime)
 
-		time.Sleep(10 * time.Microsecond)
 		sqlDB.Exec(t, `INSERT INTO foo VALUES (2, 'after')`)
+
+		// Sanity check that operations happened in the expected order.
+		require.True(t, beforeInsert.Less(insertTimestamp) && insertTimestamp.Less(tsLogical) && tsLogical.Less(s.Server.Clock().Now()),
+			fmt.Sprintf("beforeInsert: %s, insertTimestamp: %s, tsLogical: %s", beforeInsert, insertTimestamp, tsLogical))
 
 		// The below function is currently used to test negative timestamp in cursor i.e of the form
 		// "-3us".
@@ -834,7 +837,7 @@ func TestChangefeedCursor(t *testing.T) {
 		// We do not need to override for the remaining cases
 		knobs.OverrideCursor = nil
 
-		fooLogical := feed(t, f, `CREATE CHANGEFEED FOR foo WITH cursor=$1`, tsLogical)
+		fooLogical := feed(t, f, `CREATE CHANGEFEED FOR foo WITH cursor=$1`, eval.TimestampToDecimalDatum(tsLogical).String())
 		defer closeFeed(t, fooLogical)
 		assertPayloads(t, fooLogical, []string{
 			`foo: [2]->{"after": {"a": 2, "b": "after"}}`,
@@ -863,7 +866,7 @@ func TestChangefeedCursor(t *testing.T) {
 				jobutils.InternalSystemJobsBaseQuery), e.JobID()).Scan(&bytes)
 			var payload jobspb.Payload
 			require.NoError(t, protoutil.Unmarshal(bytes, &payload))
-			require.Equal(t, parseTimeToHLC(t, tsLogical), payload.GetChangefeed().StatementTime)
+			require.Equal(t, tsLogical, payload.GetChangefeed().StatementTime)
 		}
 	}
 

